strutil/strutil.go

```diff
@@ -511,3 +512,35 @@ func Reverse(in string) string {
 	}
 	return string(out)
 }
+
+// ReplaceNonMatcher replaces all characters within a given string that do not match
+// a given regular expression. The function returns the updated string and the
+// number of replacements
+func ReplaceNonMatcher(in, matcher, replaceWith string) (string, error) {
```

Comment:
It's not clear from the name/comment what this does; as the comment says it matches characters (really runes) in the given string and replaces with the given string, but someone could easily think that it's substring matching, or more specifically that it will replace all matches of the regex on the input string with the given value, without realizing that the matching will be rune-by-rune even though the replacement value is a string instead of a rune.

Possibly making this more confusing is that the user can control whether to match a single char or a string just based on the regex itself. So a user could pass in a regex matching a substring to this function, not realizing that it will never match anything because this does rune-by-rune.

My comment from before still stands -- I don't really think this function needs to exist. I think any usage of this function can be solved more flexibly by just using the stdlib `regexp.ReplaceAllString` and, if the caller needs, forming their regex to negate the character set they want. This function basically does the same thing except forces a rune-by-rune comparison, which AFAIK can be done with the regex anyways.

Reply:
Well, thanks for the feedback @jefferai. I understand that you do not be 100% aligned with that description. If you have a better idea, feel free to adjust in a way so that it is more clear.

Background of this PR is to have a base that I/people can use in a sanitazation function that we can hand out as best practice. I got adviced by multiple members of team security to place the code in this repo. I am wondering why a security related function should not take place in this repo? What would be your recommendation? I do not want to create another repository like `go-secure-stdlib2`. 

Using `regexp.ReplaceAllString` is perfect to implement a deny-list and remove single character that seems to be bogus. Implementing a deny list is a security anti-pattern. A common best practice that is handed out from security is to implement allow lists. This motivation given, we need an inverted version of `regexp.ReplaceAllString` which replaces everything that does not match a given regex. Having such a base-function in the `strutils` would make implementing of allow-lists much easier.

I highly appreciate your feedback and it super valuable.

Reply:
I think my suggestion is exactly the opposite of what you are saying re: allow/deny list: by creating a regex that explicitly says what is allowed, and then saying "deny anything that isn't on this list", by doing it the way I suggested you are actually creating an allow list and then telling it to replace anything that doesn't match.

The current mechanism you're suggesting is to do the opposite: to create a regex that must match _everything_ that isn't allowed. If that problem space expands, you will no longer filter things that should be filtered.

Again, it's not that I think a security related function shouldn't exist in this repo; it's that I don't think this function provides utility over what's in the standard library, and I think doing things the way they are generally done (via specifying the allow list in the regex) as opposed to what's suggested here is actually the more secure option.